### ISSUE TITLE
Some shader files are missing

### ISSUE THREAD

**Issue by benne3333** [vsg-dev/vsgExamples#123]
For some of the vsgExamples shader files are missing in the repo. I.e.: simple_raygen.rgen for the raytracing example (as hit & miss do).

**Comment by robertosfield** [comment]
simple_raygen.rgen exist in the repo:

https://github.com/vsg-dev/vsgExamples/blob/master/data/shaders/simple_raygen.rgen

What makes you think these files don't exist?  If you can be specific then I can look into it.  For now though I don't have anything to look into as the example you suggest does exist in the repo.

**Comment by benne3333** [comment]
Sorry, a short look made me think that they weren't copied with clone ... but they are there. I get a "Could not create shaders." when starting some of the examples (e.g. vsgraytracing) and thought that's the reason why. I set VSG_FILE_PATH and also tried to specify the full path in vsg::read_cast<vsg::ShaderStage>. I am overlooking something and will have a deeper look in it Monday.

**Comment by benne3333** [comment]
OK, can't see the reason why the shaders are not loaded ... I just get "Could not create shaders.". Debugging is complicated due to the sophisticated templating and lambdas ... the project is too complicated for my purposes and I'm giving up here. However, interesting project, but I'm looking for things in a more comprehensive structure for use at university.

**Comment by robertosfield** [comment]
On Mon, 22 Nov 2021 at 13:14, benne3333 ***@***.***> wrote:

> OK, can't see the reason why the shaders are not loaded ... I just get
> "Could not create shaders.".
>

Perhaps the error is misleading in your case.  HW support for Ray tracing
is no common so perhaps this is an issue in your case.

What OS, hardware and drivers are you using?

>

**Comment by rolandhill** [comment]
I had similar problems last night with a pull from a week or so ago.
Updated all my repos, rebuilt and now the problem examples all work. Make
sure you're on the latest version.

Roland

**Comment by benne3333** [comment]
I gave it another try, don't work.

**Comment by robertosfield** [comment]
On Tue, 23 Nov 2021 at 08:04, benne3333 ***@***.***> wrote:

> I gave it another try, don't work.
>
> We can't fix something if we can't determine how to recreate the problem.
What OS, hardware and drivers are you using?



>

**Comment by benne3333** [comment]
I'm on Ubuntu (Budgie) 21.10., Driver is NVidia 470.82.00 on a 2080 QMax. I only had few time to look in it in a short debug session, but I can't step into the library, because it is a separate project (from vsgExamples). Do you have an easy setup for this - would probably be easier for me than for you to "guess" from remote.

**Comment by robertosfield** [comment]
On Tue, 23 Nov 2021 at 09:07, benne3333 ***@***.***> wrote:

> I'm on Ubuntu (Budgie) 21.10., Driver is NVidia 470.82.00 on a 2080 QMax.
> I only had few time to look in it in a short debug session, but I can't
> step into the library, because it is a separate project (from vsgExamples).
> Do you have an easy setup for this - would probably be easier for me than
> for you to "guess" from remote.
>

I'm using Ubuntu 21.04 with 2060 with 470.82.00, so very similar to your
system.

Could you cd into vgExamples/data and then run vsgraytracing.  The shaders
searched from are in vsgExamples/data/shaders. the path use to find them is
: "shaders/simple_raygen.rgen"

I have VSG_FILE_PATH env set to vsgExamples/data.

I'm thinking better info about finding the shaders, or compiling in a
fallback.





>

**Comment by benne3333** [comment]
No, I continue to get the same message. Some path issues were my first guess, I've already tried to give the absolute path in the .cpp, didn't work either. So, I don't think, its a mere path issue. But some more information about what's happening would be good indeed.

**Comment by benne3333** [comment]
OK, I had a look in it. In the code below (part of read.cpp in module io), the endings .rgen, .rmiss and so on, are ignored. I think, it is assumed, that SPIR-V should be loaded there? Anyway, because the suffix is "wrong", reading the file fails.

`ref_ptr<Object> vsg::read(const Path& filename, ref_ptr<const Options> options)
{
    std::cout << filename;

    auto read_file = [&]() -> ref_ptr<Object> {
        if (options && !options->readerWriters.empty())
        {
            for (auto& readerWriter : options->readerWriters)
            {
                auto object = readerWriter->read(filename, options);
                if (object) return object;
            }
        }

        auto ext = vsg::lowerCaseFileExtension(filename);

        if (ext == "vsga" || ext == "vsgt" || ext == "vsgb")
        {
            VSG rw;
            return rw.read(filename, options);
        }
        else if (ext == "spv")
        {
            spirv rw;
            return rw.read(filename, options);
        }
        else
        {
            // no means of loading file
            return {};
        }
    };`

**Comment by robertosfield** [comment]
Hi

On Tue, 23 Nov 2021 at 13:42, benne3333 ***@***.***> wrote:

> OK, I had a look in it. In the code below (part of read.cpp in module io),
> the endings .rgen, .rmiss and so on, are ignored. I think, it is assumed,
> that SPIR-V should be loaded there? Anyway, because the suffix is "wrong",
> reading the file fails.
>
The GLSL shader reader is provided by vsgXchange, rather than the core
VulkanSceneGraph library.

So... from the sound of it you haven't built vsgXchange and compiled
vsgExamples against this.  The code is currently depending upon reading the
GLSL shaders. but clearly needs a better fallback for when vsgXchange
available.  I'll implement this now.



>

**Comment by robertosfield** [comment]
I have checked in a fix to vsgraytracing so that it reads the .spv variants of the shaders where the GLSL shaders can't be read, such as when vsgraytracing isn't compiled with vsgXchange:

22a866c750af633c3cabf8ec262f0649b8742bcc

I have tested vsgraytracing without vsgXchange and it works fine now.  @benne3333 could you try the latest vsgExamples master to see it now works.

**Comment by benne3333** [comment]
Confirmed.
I tried to cross check with vsgXchange, but there seems to be a problem with my assimp installation (cmake complaining about /usr/lib/include not being there - it really isn't, but other applications use libassimp-dev successfully). Maybe you want to have a look in that. I think this issue here could be closed.

**Comment by robertosfield** [comment]
On Tue, 23 Nov 2021 at 16:16, benne3333 ***@***.***> wrote:

> Confirmed.
>

Great, thanks for testing.


> I tried to cross check with vsgXchange, but there seems to be a problem
> with my assimp installation (cmake complaining about /usr/lib/include not
> being there - it really isn't, but other applications use libassimp-dev
> successfully).
>

What version of Assimp does Ubuntu 21.10 come with?

I previously tried libassimp-dev when I first tried the assimp submissions
to vsgXchange and it wasn't modern enough so ended up compiling from
source.  Assimp is a bit of a moving target so even when compiling against
source can be problematic.  We will need to try and find a path that works
reliably with assimp, perhaps even pulling in the assump source and
building a specific tag.

That's a whole other topic though.  I'll close this one now the problem
with vsgraytracing has been resolved.












>
